### PR TITLE
*: remove limiting process id for auto analyze

### PR DESCRIPTION
**Pull request by hawkingrei**
### What problem does this PR solve?

Issue Number: ref #53460

Problem Summary:

### What changed and how does it work?

- Add ```GlobalAutoAnalyzeProcessList``` to collect all auto analyze process IDs.
- Remove the restriction of having only one process ID for auto analyze and release the used process ID.

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  > 

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: pkg/executor/analyze_col_v2.go [hunk 1]
```diff
@@ -854,7 +855,9 @@ func readDataAndSendTask(ctx sessionctx.Context, handler *tableResultHandler, me
 	for {
 		failpoint.Inject("mockKillRunningV2AnalyzeJob", func() {
```
Comment (hawkingrei) [inline pkg/executor/analyze_col_v2.go:856]: Here are the tests.

File: pkg/domain/domain.go [hunk 2]
```diff
@@ -2840,9 +2840,8 @@ func (do *Domain) ReleaseConnID(connID uint64) {
 }
 
 // GetAutoAnalyzeProcID returns processID for auto analyze
-// TODO: support IDs for concurrent auto-analyze
 func (do *Domain) GetAutoAnalyzeProcID() uint64 {
```
Comment (winoros) [inline pkg/domain/domain.go:2843]: This function is not used after this patch?

File: pkg/statistics/handle/util/auto_analyze_proc_id_generator.go [hunk 3]
```diff
@@ -14,27 +14,112 @@
...
 // AutoAnalyzeProcIDGenerator is used to generate auto analyze proc ID.
 type AutoAnalyzeProcIDGenerator interface {
 	// AutoAnalyzeProcID generates an analyze ID.
 	AutoAnalyzeProcID() uint64
+	ReleaseAutoAnalyzeProcID(uint64)
 }
 
 var _ AutoAnalyzeProcIDGenerator = (*generator)(nil)
 
 type generator struct {
```
Comment (winoros) [inline pkg/statistics/handle/util/auto_analyze_proc_id_generator.go:33]: Why is it a generator? Seems that the process is generated by the `domain.NextConnID`

File: pkg/domain/domain.go (hunk 2 above)
Comment (hawkingrei) [reply pkg/domain/domain.go]: Yes, have removed.

File: pkg/statistics/handle/util/auto_analyze_proc_id_generator.go (hunk 3 above)
Comment (hawkingrei) [reply pkg/statistics/handle/util/auto_analyze_proc_id_generator.go:33]: I think it just wants to isolate the dependency on the domain and provide a container to reference the application and release of the connect id.

**Review by xhebox — APPROVED** [review]

**Review by pingyu — APPROVED** [review]
[`reservedConnAnalyze`](https://github.com/pingcap/tidb/blob/56710aeaef70943186d4e0fa7bc233b9834fe8cb/pkg/domain/domain.go#L2859) can be removed. Rest LGTM~

**Comment by hawkingrei** [comment]
> [`reservedConnAnalyze`](https://github.com/pingcap/tidb/blob/56710aeaef70943186d4e0fa7bc233b9834fe8cb/pkg/domain/domain.go#L2859) can be removed. Rest LGTM~

I have removed it. Thanks.

**Review by lance6716 — APPROVED** [review]

File: pkg/statistics/handle/util/auto_analyze_proc_id_generator.go [hunk 4]
```diff
@@ -14,27 +14,111 @@
...
+// ReleaseAutoAnalyzeProcID implements AutoAnalyzeProcIDGenerator.
+func (g *generator) ReleaseAutoAnalyzeProcID(id uint64) {
+	g.releaseAutoAnalyzeProcID(id)
+}
+
+// GlobalAutoAnalyzeProcessList is used to track the auto analyze process.
+var GlobalAutoAnalyzeProcessList = newGlobalAutoAnalyzeProcessList()
+
+type globalAutoAnalyzeProcessList struct {
+	processes map[uint64]struct{}
```
Comment (hawkingrei) [inline pkg/statistics/handle/util/auto_analyze_proc_id_generator.go:61]: We use global thread-safe map to record the running analyze task.

**Comment by hawkingrei** [comment]
@winoros PTAL

**Review by qw4990 — APPROVED** [review]

File: pkg/statistics/handle/util/auto_analyze_proc_id_generator.go (hunk 4 above)
Comment (AilinKid) [reply pkg/statistics/handle/util/auto_analyze_proc_id_generator.go:61]: why not use a `sync.map`

Comment (hawkingrei) [reply pkg/statistics/handle/util/auto_analyze_proc_id_generator.go:61]: The Map type is optimized for two common use cases: (1) when the entry for a given key is only ever written once but read many times, as in caches that only grow, or (2) when multiple goroutines read, write, and overwrite entries for disjoint sets of keys. In these two cases, use of a Map may significantly reduce lock contention compared to a Go map paired with a separate Mutex or RWMutex. 

so two common use cases are not the same as us.

**Pull request merged**